### PR TITLE
feat: open ai feature request

### PR DESCRIPTION
**Pull request by sondermanish**
## This is feature branch for Open AI integration.

This Plugin supports two commands: 
- Chat - Open AI responds back with a list of messages.
- Embeddings - Open AI responds back with list of embeddings.

### Chat Interface: 
In chat interface the user can select a model from GPT 3.5, GPT 4, and there fine-tuned versions.
Chat-screen

### Chat Response: 
chat-response

### Embedding interface: 
Embedding screen

### Strategy: 
Added Factory classes for separating implementation of different commands, interface methods are: 
- getTriggerMethod() HTTP method to use for trigger call
- getTriggerURI() url for trigger request 
- getExecutionMethod() HTTP method to user for Execution call
- getExecutionURI () Url for executionRequest
- makeRequestBody() for sending requests
- isModelCompatible  is model compatible for this command ?

Fixes: https://github.com/appsmithorg/appsmith/issues/28269

### CODE REVIEW

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+            super(config);
+        }
+
+        @Override
+        public Mono<ActionExecutionResult> executeParameterized(
+                APIConnection connection,
+                ExecuteActionDTO executeActionDTO,
+                DatasourceConfiguration datasourceConfiguration,
+                ActionConfiguration actionConfiguration) {
+            prepareConfigurationsForExecution(executeActionDTO, actionConfiguration, datasourceConfiguration);
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java:67]: Why is this method called twice, here and on L#72?

File: app/server/appsmith-plugins/pom.xml [hunk 2]
```diff
@@ -3,144 +3,267 @@
...
             <version>${testcontainers.version}</version>
+
             <scope>test</scope>
+
             <exclusions>
+
                 <exclusion>
+
                     <groupId>junit</groupId>
+
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/pom.xml:258]: What's this about? 🤔

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java (hunk 1 above)
Comment (sondermanish) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java]: Keyboard shortcuts.. smh

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java [hunk 3]
```diff
@@ -0,0 +1,17 @@
+package com.external.plugins.constants;
+
+public class OpenAIErrorMessages {
+
+    public static final String STRING_APPENDER = "%s %s";
+    public static final String EXECUTION_FAILURE = "Query failed to execute because";
+    public static final String MODEL_NOT_SELECTED = "model hasn't been selected, Please select model";
+    public static final String INPUT_NOT_CONFIGURED = "embeddings input has not been configured properly";
+    public static final String ENCODING_CONVERSION_ERROR = "embeddinngs's encoding format is not configured properly";
```
Comment (mohanarpit) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java:9]:
```suggestion
    public static final String ENCODING_CONVERSION_ERROR = "embeddings's encoding format is not configured properly";
```

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+    public static final String QUERY_NOT_CONFIGURED = "query is not configured.";
+    public static final String BAD_TEMPERATURE_CONFIGURATION =
+            "temperature value doesn't conform to contract, please enter a real number between 0 and 2";
```
Comment (mohanarpit) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java:12]:
```suggestion
            "temperature value doesn't conform to the contract. Please enter a real number between 0 and 2";
```

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+    public static final String INCORRECT_ROLE_VALUE =
+            "role value is incorrect, Choose a role value which conforms to open ai contract";
```
Comment (mohanarpit) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java:14]:
```suggestion
            "role value is incorrect. Please choose a role value which conforms to the OpenAI contract";
```

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+    public static final String INCORRECT_MESSAGE_FORMAT =
+            "messages object is not correctly configured, please provide a list of messages";
```
Comment (mohanarpit) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java:16]:
```suggestion
            "messages object is not correctly configured. Please provide a list of messages";
```

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java [hunk 7]
```diff
@@ -0,0 +1,17 @@
+package com.external.plugins.constants;
+
+public class OpenAIErrorMessages {
+
+    public static final String STRING_APPENDER = "%s %s";
+    public static final String EXECUTION_FAILURE = "Query failed to execute because";
+    public static final String MODEL_NOT_SELECTED = "model hasn't been selected, Please select model";
```
Comment (mohanarpit) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java:7]:
```suggestion
    public static final String MODEL_NOT_SELECTED = "model hasn't been selected. Please select a model";
```

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                            Object body = objectMapper.readValue(responseEntity.getBody(), Object.class);
+                            actionExecutionResult.setBody(body);
+                        } catch (IOException ex) {
+                            actionExecutionResult.setIsExecutionSuccess(false);
+                        }
+                        return Mono.just(actionExecutionResult);
+                    })
+                    .onErrorResume(error -> {
+                        errorResult.setIsExecutionSuccess(false);
+                        error.printStackTrace();
```
Comment (mohanarpit) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java:129]: Nit: Can we remove the double stack trace printing happening here & the log message below?

File: app/server/appsmith-plugins/pom.xml (hunk 2 above)
Comment (sondermanish) [reply app/server/appsmith-plugins/pom.xml:258]: these were spotless additions, not sure why this happened though

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java (hunk 8 above)
Comment (sondermanish) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java]: changed as per suggestion

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java (hunk 7 above)
Comment (sondermanish) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java:7]: changed as per suggestion

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java (hunk 6 above)
Comment (sondermanish) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java]: changed as suggested

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java (hunk 3 above)
Comment (sondermanish) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java]: changed as suggested

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java (hunk 4 above)
Comment (sondermanish) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/constants/OpenAIErrorMessages.java]: changed as suggested

**Comment by sondermanish** [comment]
Adding Test plan approved as ai-experiments team has internally tested all features

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java [hunk 9]
```diff
@@ -0,0 +1,221 @@
...
+                    });
+        }
+
+        @Override
+        public Set<String> validateDatasource(DatasourceConfiguration datasourceConfiguration) {
+            return RequestUtils.validateBearerTokenDatasource(datasourceConfiguration);
+        }
+
+        @Override
+        public Mono<TriggerResultDTO> trigger(
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java:147]: GJ on making this UQI compliant from the get-go. :)

Comment (sondermanish) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java:147]: #copy-paste

**Review by AmanAgarwal041 — APPROVED** [review]

File: app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java (hunk 9 above)
Comment (nidhi-nair) [reply app/server/appsmith-plugins/openAiPlugin/src/main/java/com/external/plugins/OpenAiPlugin.java:147]: Lol gj on picking the correct code to copy paste in that case 😛

File: app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/ApplicationPageServiceCEImpl.java [hunk 10]
```diff
@@ -1261,14 +1261,16 @@ private Mono<Application> sendApplicationPublishedEvent(
                     Integer jsFuncCount = getActionCount(pluginTypeCollectionMap, PluginType.JS);
                     Integer saasQueryCount = getActionCount(pluginTypeCollectionMap, PluginType.SAAS);
                     Integer remoteQueryCount = getActionCount(pluginTypeCollectionMap, PluginType.REMOTE);
+                    Integer aiQueryCount = getActionCount(pluginTypeCollectionMap, PluginType.AI);
 
                     extraProperties.put("dbQueryCount", dbQueryCount);
                     extraProperties.put("apiCount", apiCount);
                     extraProperties.put("jsFuncCount", jsFuncCount);
                     extraProperties.put("saasQueryCount", saasQueryCount);
                     extraProperties.put("remoteQueryCount", remoteQueryCount);
```
Comment (nidhi-nair) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/ApplicationPageServiceCEImpl.java:1270]: Also needs to be reported individually as a prop here?

Comment (sondermanish) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/services/ce/ApplicationPageServiceCEImpl.java:1270]: Facepalm: implicitly I thought I did that, noob move!!

**Review by nidhi-nair — APPROVED** [review]

**Review by dvj1988 — APPROVED** [review]

**Pull request merged**